src/main/java/org/opensearch/flowframework/transport/ReprovisionWorkflowTransportAction.java

```diff
@@ -0,0 +1,300 @@
+     * @param flowFrameworkSettings Whether this API is enabled
+     * @param encryptorUtils Utility class to handle encryption/decryption
+     * @param pluginsService The Plugins Service
+     */
+    @Inject
+    public ReprovisionWorkflowTransportAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        ThreadPool threadPool,
+        Client client,
+        WorkflowStepFactory workflowStepFactory,
+        WorkflowProcessSorter workflowProcessSorter,
+        FlowFrameworkIndicesHandler flowFrameworkIndicesHandler,
+        FlowFrameworkSettings flowFrameworkSettings,
+        EncryptorUtils encryptorUtils,
+        PluginsService pluginsService
+    ) {
+        super(ReprovisionWorkflowAction.NAME, transportService, actionFilters, ReprovisionWorkflowRequest::new);
+        this.threadPool = threadPool;
+        this.client = client;
+        this.workflowStepFactory = workflowStepFactory;
+        this.workflowProcessSorter = workflowProcessSorter;
+        this.flowFrameworkIndicesHandler = flowFrameworkIndicesHandler;
+        this.flowFrameworkSettings = flowFrameworkSettings;
+        this.encryptorUtils = encryptorUtils;
+        this.pluginsService = pluginsService;
+    }
+
+    @Override
+    protected void doExecute(Task task, ReprovisionWorkflowRequest request, ActionListener<WorkflowResponse> listener) {
+
+        String workflowId = request.getWorkflowId();
+
+        // Original template is retrieved from index, attempt to decrypt any exisiting credentials before processing
+        Template originalTemplate = encryptorUtils.decryptTemplateCredentials(request.getOriginalTemplate());
+        Template updatedTemplate = request.getUpdatedTemplate();
+
+        // Validate updated template prior to execution
+        Workflow provisionWorkflow = updatedTemplate.workflows().get(PROVISION_WORKFLOW);
+        List<ProcessNode> updatedProcessSequence = workflowProcessSorter.sortProcessNodes(
+            provisionWorkflow,
+            request.getWorkflowId(),
+            Collections.emptyMap() // TODO : Add suport to reprovision substitution templates
+        );
+
+        try {
+            workflowProcessSorter.validate(updatedProcessSequence, pluginsService);
+        } catch (Exception e) {
+            String errormessage = "Workflow validation failed for workflow " + request.getWorkflowId();
+            listener.onFailure(new FlowFrameworkException(errormessage, RestStatus.BAD_REQUEST));
+        }
+
+        // Retrieve resources created
+        GetWorkflowStateRequest getStateRequest = new GetWorkflowStateRequest(workflowId, true);
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+            logger.info("Querying state for workflow: {}", workflowId);
+            client.execute(GetWorkflowStateAction.INSTANCE, getStateRequest, ActionListener.wrap(response -> {
+                context.restore();
+
```

Comment:
+1 I was going to mention it can be done here or earlier before we even get here. This is a good place to do it too

Reply:
I will place the check within the create workflow transport action since we need to block regular document updates if the template has a status of not started

Reply:
> since we need to block regular document updates if the template has a status of not started

what do you mean by this?

Reply:
disregard the earlier comment, I see that the progress state is checked when attempting to update the template. I will add the progress check for the reprovision transport action and block if in progress or not started